### PR TITLE
Basic syntax of 'INSERT INTO ... SELECT' statement was added (#8293).

### PR DESCRIPTION
**Pull request by voroks**
Method "insert()" accepts instance of \yii\db\Query for second parameter as well as array.
Column names (or aliases, if exists) of SELECT statement became column names for INSERT statement. So, support of basic usage of "INSERT INTO ... SELECT" statement (as in issue #8293) was added.

### CODE REVIEW

**Comment by samdark** [comment]
How about non-MySQL?

**Comment by voroks** [comment]
PostgreSQL - OK
SQLite - OK
Unfortunately, I can't perform test with using other DBs.

File: framework/db/Command.php [hunk 1]
```diff
@@ -418,7 +418,8 @@ public function queryColumn()
      * Note that the created command is not executed until [[execute()]] is called.
      *
      * @param string $table the table that new rows will be inserted into.
-     * @param array $columns the column data (name => value) to be inserted into the table.
+     * @param array|\yii\db\Query $columns the column data (name => value) to be inserted into the table or instance
+     * of \yii\db\Query to perform INSERT INTO ... SELECT SQL statement.
```
Comment (cebe) [inline framework/db/Command.php]: Need a note that the `Query` can be passed since version 2.0.11.

File: framework/db/QueryBuilder.php [hunk 2]
```diff
@@ -123,7 +123,8 @@ public function build($query, $params = [])
      * The method will properly escape the table and column names.
      *
      * @param string $table the table that new rows will be inserted into.
-     * @param array $columns the column data (name => value) to be inserted into the table.
+     * @param array|\yii\db\Query $columns the column data (name => value) to be inserted into the table or instance
+     * of \yii\db\Query to perform INSERT INTO ... SELECT SQL statement.
```
Comment (cebe) [inline framework/db/QueryBuilder.php]: Need a note that the Query can be passed since version 2.0.11.

**Review by cebe — CHANGES_REQUESTED** [review]
test seems to pass on travis so this looks good to me. A CHANGELOG is needed and it should be rebased on current master.

**Comment by cebe** [comment]
@voroks are you willing to work on this further?

File: framework/db/mysql/QueryBuilder.php [hunk 3]
```diff
@@ -195,30 +195,42 @@ public function insert($table, $columns, &$params)
         }
         $names = [];
         $placeholders = [];
-        foreach ($columns as $name => $value) {
-            $names[] = $schema->quoteColumnName($name);
```
Comment (cebe) [inline framework/db/mysql/QueryBuilder.php]: no need to check `=== true`. Change this to:

```php
if ($columns instanceof \yii\db\Query) {
```

File: framework/db/oci/QueryBuilder.php [hunk 4]
```diff
@@ -178,30 +178,42 @@ public function insert($table, $columns, &$params)
         }
         $names = [];
         $placeholders = [];
-        foreach ($columns as $name => $value) {
-            $names[] = $schema->quoteColumnName($name);
```
Comment (cebe) [inline framework/db/oci/QueryBuilder.php]: no need to check `=== true`. Change this to:

```php
if ($columns instanceof \yii\db\Query) {
```

File: framework/db/pgsql/QueryBuilder.php [hunk 5]
```diff
@@ -245,17 +245,21 @@ public function update($table, $columns, $condition, &$params)
...
-     * @param array $columns the column data (name => value) to be saved into the table.
+     * @param array|\yii\db\Query $columns the column data (name => value) to be saved into the table or instance
+     * of \yii\db\Query to perform INSERT INTO ... SELECT SQL statement.
+     * Passing of \yii\db\Query is available since version 2.0.11.
      * @return array normalized columns
      * @since 2.0.9
      */
     private function normalizeTableRowData($table, $columns)
     {
-        if (($tableSchema = $this->db->getSchema()->getTableSchema($table)) !== null) {
```
Comment (cebe) [inline framework/db/pgsql/QueryBuilder.php]: no need to check `=== false`. Also better not wrap the whole function in the if but return early, so change this to:

```php
if ($columns instanceof \yii\db\Query) {
    return $columns;
}
```

**Review by cebe — CHANGES_REQUESTED** [review]

File: tests/framework/db/CommandTest.php [hunk 6]
```diff
@@ -321,6 +321,51 @@ public function testInsert()
...
+                'name' => 'test',
+                'address' => 'test address',
+            ]
+        )->execute();
+
+	$query = new \yii\db\Query();
+	$query->select([
+                '{{customer}}.email as name',
+                'name as email',
+                'address',
```
Comment (cebe) [inline tests/framework/db/CommandTest.php]: could you add the alias syntax `'alias' => 'column'`. Seems this is not supported by your implementation right now.

> When the columns are specified as an array, you may also use array keys as the column aliases (if a column
>  does not need alias, do not use a string key).
> http://www.yiiframework.com/doc-2.0/yii-db-query.html#select%28%29-detail

sorry, have not noticed this earlier.

**Review by cebe — CHANGES_REQUESTED** [review]

File: framework/db/QueryBuilder.php [hunk 7]
```diff
@@ -157,23 +159,56 @@ public function insert($table, $columns, &$params)
...
-            . (!empty($placeholders) ? ' VALUES (' . implode(', ', $placeholders) . ')' : ' DEFAULT VALUES');
+            . (!empty($placeholders) ? ' VALUES (' . implode(', ', $placeholders) . ')' : $values);
+    }
+
+    /**
+     * Prepare select-subquery and field names for INSERT INTO ... SELECT SQL statement.
+     *
+     * @param \yii\db\Query $columns Object, which represents select query
+     * @param \yii\db\Schema $schema Schema object to qoute column name
+     * @return array
```
Comment (samdark) [inline framework/db/QueryBuilder.php]: Need `@since 2.0.11`

**Review by samdark — APPROVED** [review]

File: framework/db/QueryBuilder.php [hunk 8]
```diff
@@ -157,23 +159,58 @@ public function insert($table, $columns, &$params)
...
-            $names[] = $schema->quoteColumnName($name);
-            if ($value instanceof Expression) {
-                $placeholders[] = $value->expression;
-                foreach ($value->params as $n => $v) {
-                    $params[$n] = $v;
+        $values = ' DEFAULT VALUES';
+        if ($columns instanceof \yii\db\Query) {
+            list($names, $values) = $this->prepareInsertSelectSubQuery($columns, $schema);
+        } else {
+            foreach ($columns as $name => $value) {
```
Comment (SilverFire) [inline framework/db/QueryBuilder.php]: @yiisoft/core-developers how about moving this part into a separate method to make `insert` simpler to read?

Comment (samdark) [inline framework/db/QueryBuilder.php]: I'm OK with that.

Comment (dynasource) [inline framework/db/QueryBuilder.php]: agree. Placeholders are not always defined. 
I would expect a separate method for the creation of an "insert" statement with all the relevant arguments.

File: framework/db/QueryBuilder.php [hunk 9]
```diff
@@ -157,23 +159,58 @@ public function insert($table, $columns, &$params)
...
-                $placeholders[] = $phName;
-                $params[$phName] = !is_array($value) && isset($columnSchemas[$name]) ? $columnSchemas[$name]->dbTypecast($value) : $value;
             }
         }
 
         return 'INSERT INTO ' . $schema->quoteTableName($table)
             . (!empty($names) ? ' (' . implode(', ', $names) . ')' : '')
-            . (!empty($placeholders) ? ' VALUES (' . implode(', ', $placeholders) . ')' : ' DEFAULT VALUES');
+            . (!empty($placeholders) ? ' VALUES (' . implode(', ', $placeholders) . ')' : $values);
+    }
```
Comment (dynasource) [inline framework/db/QueryBuilder.php]: the creation of the insert statement could be done a lot cleaner in a separate method returning a <<<SQL block

File: framework/db/oci/QueryBuilder.php [hunk 10]
```diff
@@ -178,30 +178,35 @@ public function insert($table, $columns, &$params)
...
+                    $names[] = $schema->quoteColumnName($name);
+                    $placeholders[] = 'DEFAULT';
+                }
             }
         }
 
         return 'INSERT INTO ' . $schema->quoteTableName($table)
             . (!empty($names) ? ' (' . implode(', ', $names) . ')' : '')
-            . (!empty($placeholders) ? ' VALUES (' . implode(', ', $placeholders) . ')' : ' DEFAULT VALUES');
+            . (!empty($placeholders) ? ' VALUES (' . implode(', ', $placeholders) . ')' : $values);
```
Comment (dynasource) [inline framework/db/oci/QueryBuilder.php]: this should be refactored to a solution in which with https://github.com/yiisoft/yii2/pull/9420/files#diff-e2271b5ca7ffbb33d4f2bfa5ed2e94f4R181 is reused

File: tests/framework/db/CommandTest.php [hunk 11]
```diff
@@ -321,6 +321,136 @@ public function testInsert()
         ], $record);
     }
 
+    /**
+     * Test INSERT INTO ... SELECT SQL statement
+     */
+    public function testInsertSelect()
+    {
```
Comment (dynasource) [inline tests/framework/db/CommandTest.php]: please split these tests into separate selfcontained methods

**Review by dynasource — CHANGES_REQUESTED** [review]

**Comment by samdark** [comment]
Merged https://github.com/yiisoft/yii2/commit/25f08afc96035b517d6a7be13b28bb09514bfc39. Thank you!

**Pull request closed without merging**